### ISSUE TITLE
Error when training using default_backend: {EXLA.Backend, []}

### ISSUE THREAD

**Issue by hiramegl** [acalejos/exgboost#40]
Hi again Andrés,

I managed to run this example from the documentation: https://hexdocs.pm/exgboost/iris_classification.html
In this Iris example there is no EXLA.Backend configured and executes correctly and fast since is a small dataset.

The problem for me now is that I'm trying to train a new model with much more data (300,000 rows x 30 cols, floats).
If I don't use the EXLA.Backend the tensor operations run very slow.
They take many minutes to finish but I manage to train the model.
Afterwards, when trying to use "EXGBoost.predict" the whole notebook crashes and I cannot use the model I trained.

When using default_backend: {EXLA.Backend, []} I get the following error while training:

booster_ms =
  EXGBoost.train(x_train, y_train,
    num_class: 2,
    objective: :multi_softmax,
    num_boost_rounds: 10000,
    evals: [{x_train, y_train, "training"}]
  )

** (FunctionClauseError) no function clause matching in Jason.Encoder.Nx.Tensor.encode/2    
    
    The following arguments were given to Jason.Encoder.Nx.Tensor.encode/2:
    
        # 1
        #Nx.Tensor<
          f32
          EXLA.Backend<host:0, 0.904935439.2317484067.245192>
          NaN
        >
    
        # 2
        {#Function<3.65781566/3 in Jason.Encode.escape_json>,
         #Function<1.65781566/3 in Jason.Encode.map_naive>}
    
    Attempted function clauses (showing 4 out of 4):
    
        def encode(%Nx.Tensor{data: %Nx.BinaryBackend{state: <<32704::integer-native-size(16)>>}}, _opts)
        def encode(%Nx.Tensor{data: %Nx.BinaryBackend{state: <<32256::integer-native-size(16)>>}}, _opts)
        def encode(%Nx.Tensor{data: %Nx.BinaryBackend{state: <<2_143_289_344::integer-native-size(32)>>}}, _opts)
        def encode(%Nx.Tensor{data: %Nx.BinaryBackend{state: <<9_221_120_237_041_090_560::integer-native-size(64)>>}}, _opts)
    
    (exgboost 0.5.0) lib/exgboost/internal.ex:114: Jason.Encoder.Nx.Tensor.encode/2
    (jason 1.4.1) lib/encode.ex:188: Jason.Encode.map_naive/3
    (jason 1.4.1) lib/encode.ex:37: Jason.Encode.encode/2
    (jason 1.4.1) lib/jason.ex:162: Jason.encode!/2
    (exgboost 0.5.0) lib/exgboost/dmatrix.ex:310: EXGBoost.DMatrix.from_tensor/2
    (exgboost 0.5.0) lib/exgboost.ex:249: EXGBoost.train/3
    #cell:bbyqcg72622y2xj7kieq53l4dlovyxli:2: (file)

I dont' know If I still need to install something else or the configuration is wrong.

Mix.install(
  [
    {:exgboost, "~> 0.5"},
    {:axon, "~> 0.6"},
    {:exla, "~> 0.6.4"},
    {:scidata, "~> 0.1"},
    {:scholar, "~> 0.2.1"},
    {:explorer, "~> 0.8.0"},
    {:kino, "~> 0.12.3"},
    {:kino_explorer, "~> 0.1.18"},

  ],
  config: [
    nx: [
      default_defn_options: [compiler: EXLA],
      default_backend: {EXLA.Backend, []}
    ]
  ]
)

I'm using the dataset from https://www.kaggle.com/datasets/nelgiriyewithana/credit-card-fraud-detection-dataset-2023

Thanks again for your help beforehand.

Best regards,

/Hiram Galicia
[CreditCardFraud.livemd.md](https://github.com/acalejos/exgboost/files/14352249/CreditCardFraud.livemd.md)

**Comment by acalejos** [comment]
Hey, glad you're taking time to explore the various notebooks and apologies for these quirks. Im actually imrpoving the docs at the moment regarding the issue you've encountered with EXLA. The solution is outlined in #22 . I also just (within the past day or 2) updated the docs to reflect this limitation (https://github.com/acalejos/exgboost?tab=readme-ov-file#known-limitations). 

It is something that can be fixed, but the fix I've determined so far isn't ideal since I'd prefer to not require a backend copy, but that might be the solution we have to go with. 

Regarding the limit as your dataset increases, that's also something that needs clarification in the docs, but we had an issue earlier that might be related (#23).

For that, try what I outlined here: https://github.com/acalejos/exgboost/issues/23#issuecomment-1722663846

Let me know if these don't address your needs, thanks. 

So basically to do what you need to do, try increasing the dirty NIF scheduler stack size and backend copying to the binary backend before predictions

**Comment by hiramegl** [comment]
Hi Andrés,

Thanks for your answer!
I will test those tips and get back later.

Best regards,

/Hiram

**Comment by hiramegl** [comment]
Hi again Andrés,

It seems that it did not help to make a backend copy of the tensors.
I've been digging a bit in the "0.5.0" code and it seems that the process is crashing when trying to run Jason.encode in the configuration map:

Tag: 0.5.0
exgboost/lib/exboost.ex: 249
249    dmat = DMatrix.from_tensor(x, y, Keyword.put_new(dmat_opts, :format, :dense))

 exgboost/lib/exgboost/dmatrix.ex: 310
304    config = Enum.into(config_opts, %{}, fn {key, value} -> {Atom.to_string(key), value} end)
305    format = Keyword.fetch!(format_opts, :format)
306
307    dmat =
308      EXGBoost.NIF.dmatrix_create_from_dense(
309        Jason.encode!(ArrayInterface.from_tensor(tensor)),
310        Jason.encode!(config) <----------------------------------- CRASHES HERE
311      )
312      |> Internal.unwrap!()

For some reason it seems that the "config" has a NaN value that the encoder cannot process.
What I did was to change the backend just before executing "train()", like this:

Nx.global_default_backend(EXLA.Backend)
... # run code that users EXLA tensors and executes fast
Nx.global_default_backend(Nx.BinaryBackend)
EXGBoost.train(x_train, y_train, ...)

x_train and y_train are still EXLA.Backend tensors, from what I can see, but the "train()" function runs without crashing.
That solves the error with "train()".

The second problem with "EXGBoost.predict()" when the whole notebook crashes I haven't been able to solve it with the flag.
I actually don't know how to add the configuration flag "+sssdcpu 128" since I'm using the Livebook installation for macos, not running from docker or source code.

I found a file "/Applications/Livebook.app/Contents/Resources/rel/releases/0.11.4" with name "vm.args" and added the configuration flags there, but the change did not made any difference, the "predict()" method still crashes. I've posted a question in Elixir forum to check if is there any way to add this flag to Livebook. Hope somebody knows :-)

Anyway, I ended up batching the evaluation tensor, converting into list and then running "predict" in the smaller tensors.
Using batches of 40_000 rows made the trick.

Thanks a lot for your help Andrés and I really appreciated your prompt answers!

Best regards,

/Hiram Galicia

**Comment by hiramegl** [comment]
Got the answer now from Elixir forum.
It seems I just needed to create a file ~/.livebookdesktop.sh with contents:
export ELIXIR_ERL_OPTIONS="+sssdcpu 512"

I only had to increase the suggested stack size to 512 instead of 128.
Then it worked like a charm.

Thanks for the help!

Best regards,

/Hiram Galicia

**Comment by acalejos** [comment]
@hiramegl Glad that you were able to find what you needed on Elixir Forum. I assume that only solved the crashing with large data issue? Did you also have to backend transfer to get the encoding to work?

Right now the Jason implementation is only done for Nx tensors on the binary backend, which is why it breaks for EXLA, and why a backend transfer to the Binary backend is the current workaround.

That's a good tidbit about adding stack size in Livebook and probably worth adding to the docs here. 

I will go ahead and close this.  If you're still experiencing issues please don't hesitate to comment on here and I can reopen.
